src/providers/mdal/qgsmdalprovider.cpp

```diff
@@ -241,6 +234,39 @@ bool QgsMdalProvider::persistDatasetGroup( const QString &path,
   return false;
 }
 
+void QgsMdalProvider::loadData()
+{
+  QByteArray curi = dataSourceUri().toUtf8();
+  mMeshH = MDAL_LoadMesh( curi.constData() );
+  if ( mMeshH )
+  {
+    const QString proj = MDAL_M_projection( mMeshH );
+    if ( !proj.isEmpty() )
+      mCrs.createFromString( proj );
+  }
+}
+
+void QgsMdalProvider::reloadData()
+{
+  if ( mMeshH )
+    MDAL_CloseMesh( mMeshH );
+
+  loadData();
+
+  if ( mMeshH )
+    for ( auto uri : mExtraDatasetUris )
+    {
+      std::string str = uri.toStdString();
+      MDAL_M_LoadDatasets( mMeshH, str.c_str() );
+
+      if ( MDAL_LastStatus() != MDAL_Status::None )
+      {
+        mExtraDatasetUris.removeOne( uri );
+      }
+    }
+
```

Comment:
I think you need to emit `QgsMeshDataProvider::dataChanged` signal here to notify GUI to reload widgets and refresh canvas. 

What I am not sure is about removal of extra datasets. So far we have just signal `datasetGroupsAdded`, but maybe we need also `datasetGroupsRemoved` to emit here to notify that we removed X datasets so the dataset group tree can be properly refreshed too.

Can you please try

1. load mesh with extra dataset (so bed elevation + extra one)
2. change mesh so extra dataset is not valid anymore
3. check the mesh layer properties to see if you have only bed elevation dataset loaded

Reply:
> I think you need to emit `QgsMeshDataProvider::dataChanged` signal here to notify GUI to reload widgets and refresh canvas.

There is a problem :
QgsDataProvider::dataChanged is connect with refreshMapCanvas
```suggestion
https://github.com/qgis/QGIS/blob/4aaa523977c9ec74ddb2e92f704eaa30d16d5121/src/app/qgisapp.cpp#L12346-L12350
```
This method call QgsMapCanvas::refreshAllLAyer (line 9401 same file)
```suggestion
https://github.com/qgis/QGIS/blob/4aaa523977c9ec74ddb2e92f704eaa30d16d5121/src/app/qgisapp.cpp#L9394-L9401
```
QgsMapCanvas::refreshAllLayer call QgsMapLayer::reload :
```suggestion
https://github.com/qgis/QGIS/blob/4aaa523977c9ec74ddb2e92f704eaa30d16d5121/src/gui/qgsmapcanvas.cpp#L2188-L2194
```
and QgsMeshLayer::reload() call QgsMeshDataProvider::reloadData() making a **circular reference**

Is it really necessary to emit QgsMeshDataProvider::dataChanged from here ? 
May the caller of QgsMeshLayer::reload() or QgsMeshDataProvider::reloadData() can keep the responsability to emit dataChanged ?

Reply:
> Can you please try
> 
>     1. load mesh with extra dataset (so bed elevation + extra one)
> 
>     2. change mesh so extra dataset is not valid anymore
> 
>     3. check the mesh layer properties to see if you have only bed elevation dataset loaded

Something similar has been done in the added test, no ?

Reply:
> I think you need to emit `QgsMeshDataProvider::dataChanged` signal here to notify GUI to reload widgets and refresh canvas.

There is a problem :
`QgsDataProvider::dataChanged` is connect with `refreshMapCanvas` (line 12349 in qgisapp.cpp)

This method calls `QgsMapCanvas::refreshAllLayers` (line 9401 same file)

`QgsMapCanvas::refreshAllLayer()` calls `QgsMapLayer::reload() `

and `QgsMeshLayer::reload()` calls `QgsMeshDataProvider::reloadData()` 

making a **circular reference**

Is it really necessary to emit `QgsMeshDataProvider::dataChanged` from here? 
May the caller of `QgsMeshLayer::reload()` or `QgsMeshDataProvider::reloadData()` can keep the responsability to inform the data has changed when necessary ?

Reply:
> What I am not sure is about removal of extra datasets. So far we have just signal `datasetGroupsAdded`, but maybe we need also `datasetGroupsRemoved` to emit here to notify that we removed X datasets so the dataset group tree can be properly refreshed too.

Remove extra dataSets is not as simple it looks like... I look at the code of the class  `QgsMeshRendererSettings` and the renderer settings are stored in a QHash with index of the groupsDataset as key. If we remove extra dataset, the index will change... Maybe  we need to store the renderer settings in QList to make easier the synchronization with the `mExtraDatasetUris` list.
It make me thinking about the solution I have purpose sooner: not remove the datasetGroup if it is no valide or incompatible but replace it by a void datasetGroup. So no need to delete the datasetGroup...
On the other hand, It is interesting to open the possibility to remove extra datasets : the user may wants to remove extra datasets he doesn't want its anymore.
What do you think about ?

Reply:
ok with `dataChanged()`. we just need to make sure that this `reload()` is not called with every refresh of the canvas, which is not the case (Recreating the mesh/triangular mesh can take ages for large datasets)

as for `mExtraDatasetUris` it is even more complicated because one extra URI can have multiple dataset groups in it (e.g. XMDF format or so).

it would be good to have possibility to remove datasets from gui, but that would need some work in MDAL too (to have some kind of UUID of datasets or so) so we can persist in QGIS project which datasets should not be loaded. 

for now I would just simply add there another function similar to `QgsMeshLayer::onDatasetGroupsAdded` which would just reassign all `assignDefaultStyleToDatasetGroup` to all dataset groups. But I would do this ONLY if we had to remove some extra uris on the way (because it would essentially remove all user settings).

Reply:
Hi,
ok, I will work on this signal/slot.

But just an idea : if the difficulty comes from extra URI which can have multiple dataset groups, what about a table who store all the dataset groups that are needed to be loaded (valid, no valid, compatible, no compatible,...) ? This table could store information which could be used to handle the dataset groups.
The entry key could be the index group for Qgis and the stored information could be the uri, the index for MDAL if any (if valid and compatible),... . Access method of the provider will used this table to retrieve the MDAL corresponding dataset group and return it, or return void dataset group if no valid or not compatible. This table could be refresh when reload is called.
This table could be the base for a table model associate to a table view in the mesh properties dialog.
Now, it is just an idea... but it could be not complicated to implement.

Reply:
whatever works :) but the MDAL group index is tricky, since if you load extraUris in different order it is different index. We would probably need that global id for groups some time anyways.

Reply:
ok, done.
I have added a way to avoid that non extras dataset groups have user setting removed.

I look on the MDAL code, and it seems the extra dataset groups are always added at the end of the DatasetGroups So, I don't understand why MDAL group index is tricky? The provider load extraUris always in the same order, so the the dataset groups have always the same order, no ?

Reply:
yes